mapo/agents/off_mapo/off_mapo_policy.py

```diff
@@ -0,0 +1,372 @@
+        self.evaluation = evaluate
+
+
+class TargetUpdatesMixin:  # pylint: disable=too-few-public-methods
+    """Adds methods to build ops that update target networks."""
+
+    def build_update_targets_op(self, tau=None):
+        """Build op to update target networks."""
+        tau = tau or self.config["tau"]
+        main_variables = self.main_variables
+        target_variables = self.target_variables
+        update_target_expr = [
+            target_var.assign(tau * var + (1.0 - tau) * target_var)
+            for var, target_var in zip(main_variables, target_variables)
+        ]
+        return tf.group(*update_target_expr)
+
+
+class DelayedUpdatesMixin:  # pylint: disable=too-few-public-methods
+    """
+    Sets ops to update models with different frequencies.
+
+    This mixin is needed since `build_tf_policy` does not have an argument to
+    override `TFPolicy.build_apply_op`.
+    """
+
+    @override(TFPolicy)
+    def build_apply_op(self, *_):
+        """
+        Update actor and critic models with different frequencies.
+
+        For policy gradient, update policy net one time v.s. update critic net
+        `policy_delay` time(s). Also use `policy_delay` for target networks update.
+        """
+        with tf.control_dependencies([self.global_step.assign_add(1)]):
+            # Critic updates
+            critic_op = self._critic_optimizer.apply_gradients(
+                self._critic_grads_and_vars
+            )
+            # Actor updates
+            should_apply_actor_opt = tf.equal(
+                tf.math.mod(self.global_step, self.config["policy_delay"]), 0
+            )
+
+            def make_actor_apply_op():
+                return self._actor_optimizer.apply_gradients(self._actor_grads_and_vars)
+
+            with tf.control_dependencies([critic_op]):
+                actor_op = tf.cond(
+                    should_apply_actor_opt,
+                    true_fn=make_actor_apply_op,
+                    false_fn=tf.no_op,
+                )
+            # increment global step & apply ops
+            apply_ops = tf.group(actor_op, critic_op)
+            with tf.control_dependencies([apply_ops]):
+                update_targets_op = tf.cond(
+                    should_apply_actor_opt,
+                    true_fn=self.build_update_targets_op,
+                    false_fn=tf.no_op,
+                )
+            return tf.group(apply_ops, update_targets_op)
+
+
+OffMAPOTFPolicy = build_tf_policy(
+    name="OffMAPOTFPolicy",
+    loss_fn=build_actor_critic_losses,
+    get_default_config=get_default_config,
+    postprocess_fn=ignore_timeout_termination,
+    stats_fn=extra_loss_fetches,
+    optimizer_fn=lambda *_: None,
+    gradients_fn=actor_critic_gradients,
+    extra_action_feed_fn=extra_action_feed_fn,
+    before_init=setup_early_mixins,
+    before_loss_init=create_separate_optimizers,
+    after_init=copy_targets,
+    make_action_sampler=build_actor_critic_models,
```

Comment:
@angelolovatto I'm going to close this PR regardless of this minor issue, but it should keep this in mind in the future... ok?

Reply:
The documentation in updated to version 0.7.3 of Ray, whereas we were using 0.7.2 at the time of writing this code.

I've set `setup.py` to require `"ray[rllib]==0.7.2"` and opened #29 so that we can address incompatibilities and upgrade the code later.